library(ape)


generate.xml.file <- function(fasta.alignment.path, outfile.name, xml.type = "real"){
    fasta.alignment <- read.dna(fasta.alignment.path, format = "fasta")
    file.name <- outfile.name

#################################################
#################################################
block.1 <- "<?xml version=\"1.0\" standalone=\"yes\"?>
<!-- Generated by BEAUTi v1.7.5                                              -->
<!--       by Alexei J. Drummond, Andrew Rambaut and Marc A. Suchard         -->
<!--       Department of Computer Science, University of Auckland and        -->
<!--       Institute of Evolutionary Biology, University of Edinburgh        -->
<!--       David Geffen School of Medicine, University of California, Los Angeles-->
<!--       http://beast.bio.ed.ac.uk/                                        -->
<beast>
	<!-- The list of taxa to be analysed (can also include dates/ages).          -->
	<!-- ntax=10                                                                 -->
	<taxa id=\"taxa\">"

#taxa block for real dates
######################################################################
get.real.dates <- function(x){
    block.2 <- ""
    for(i in 1:length(rownames(fasta.alignment))){
        tax.name <- rownames(fasta.alignment)[i]
        t.name <- paste0("<taxon id=\"", tax.name, "\">")
        t.date <- strsplit(tax.name, "_")[[1]][2]
        t.block <- paste0("<date value=\"", t.date, "\" direction=\"forwards\" units=\"years\"/>")
        t.close <- "</taxon>"
        block.2 <- c(block.2, t.name, t.block, t.close)
    }
    block.2 <- c(block.2, "</taxa>")
    return(block.2)
}
##########################################################################

#taxa block for randomised data
##########################################################################
get.random.dates <- function(x){
    block.2.random <- ""

    taxa.dates <- vector()
    for(i in 1:nrow(fasta.alignment)){
        tax.name <- rownames(fasta.alignment)[i]
        t.date <- strsplit(tax.name, "_")[[1]][2]
        t.block <- paste0("<date value=\"", t.date, "\" direction=\"forwards\" units=\"years\"/>")
        taxa.dates <- c(taxa.dates, t.block)
    }
	taxa.dates <- taxa.dates[sample(1:nrow(fasta.alignment), nrow(fasta.alignment))]
    block.2.random <- vector()
    for(i in 1:nrow(fasta.alignment)){
        tax.name <- rownames(fasta.alignment)[i]
        t.name <- paste0("<taxon id=\"", tax.name, "\">")
        t2.block <- taxa.dates[i]
        #taxa.dates <- taxa.dates[taxa.dates != t.block]
        t.close <- "</taxon>"
        block.2.random <- c(block.2.random, t.name, t2.block, t.close)
    }
    block.2.random <- c(block.2.random, "</taxa>")
    return(block.2.random)
}
###########################################################################

block.3 <- "<alignment id=\"alignment\" dataType=\"nucleotide\">"
for(i in 1:nrow(fasta.alignment)){
    open.tax <- "<sequence>"
    tax.name <- paste0("<taxon idref=\"", rownames(fasta.alignment)[i], "\"/>")
    seq.text <- paste(as.character(fasta.alignment[i, ]), collapse = "")
    close.tax <- "</sequence>"
    block.3 <- c(block.3, open.tax, tax.name, seq.text, close.tax)
}
block.3 <- c(block.3, "</alignment>")

block.4 <- "<patterns id=\"patterns\" from=\"1\" strip=\"false\">
		<alignment idref=\"alignment\"/>
	</patterns>
	<constantSize id=\"constant\" units=\"years\">
		<populationSize>
			<parameter id=\"constant.popSize\" value=\"860.0\" lower=\"0.0\"/>
		</populationSize>
	</constantSize>
	<coalescentSimulator id=\"startingTree\">
		<taxa idref=\"taxa\"/>
		<constantSize idref=\"constant\"/>
	</coalescentSimulator>
	<treeModel id=\"treeModel\">
		<coalescentTree idref=\"startingTree\"/>
		<rootHeight>
			<parameter id=\"treeModel.rootHeight\"/>
		</rootHeight>
		<nodeHeights internalNodes=\"true\">
			<parameter id=\"treeModel.internalNodeHeights\"/>
		</nodeHeights>
		<nodeHeights internalNodes=\"true\" rootNode=\"true\">
			<parameter id=\"treeModel.allInternalNodeHeights\"/>
		</nodeHeights>
	</treeModel>"


block.5 <- "	<coalescentLikelihood id=\"coalescent\">
		<model>
			<constantSize idref=\"constant\"/>
		</model>
		<populationTree>
			<treeModel idref=\"treeModel\"/>
		</populationTree>
	</coalescentLikelihood>

	<!-- The uncorrelated relaxed clock (Drummond, Ho, Phillips & Rambaut (2006) PLoS Biology 4, e88 )-->
	<discretizedBranchRates id=\"branchRates\">
		<treeModel idref=\"treeModel\"/>
		<distribution>
			<logNormalDistributionModel meanInRealSpace=\"true\">
				<mean>
					<parameter id=\"ucld.mean\" value=\"1.0\" lower=\"0.0\"/>
				</mean>
				<stdev>
					<parameter id=\"ucld.stdev\" value=\"0.3333333333333333\" lower=\"0.0\"/>
				</stdev>
			</logNormalDistributionModel>
		</distribution>
		<rateCategories>
			<parameter id=\"branchRates.categories\"/>
		</rateCategories>
	</discretizedBranchRates>
	<rateStatistic id=\"meanRate\" name=\"meanRate\" mode=\"mean\" internal=\"true\" external=\"true\">
		<treeModel idref=\"treeModel\"/>
		<discretizedBranchRates idref=\"branchRates\"/>
	</rateStatistic>
	<rateStatistic id=\"coefficientOfVariation\" name=\"coefficientOfVariation\" mode=\"coefficientOfVariation\" internal=\"true\" external=\"true\">
		<treeModel idref=\"treeModel\"/>
		<discretizedBranchRates idref=\"branchRates\"/>
	</rateStatistic>
	<rateCovarianceStatistic id=\"covariance\" name=\"covariance\">
		<treeModel idref=\"treeModel\"/>
		<discretizedBranchRates idref=\"branchRates\"/>
	</rateCovarianceStatistic>"
	

block.6 <- "<gtrModel id=\"gtr\">
		<frequencies>
			<frequencyModel dataType=\"nucleotide\">
				<frequencies>
					<parameter id=\"frequencies\" value=\"0.25 0.25 0.25 0.25\"/>
				</frequencies>
			</frequencyModel>
		</frequencies>
		<rateAC>
			<parameter id=\"ac\" value=\"1.0\" lower=\"0.0\"/>
		</rateAC>
		<rateAG>
			<parameter id=\"ag\" value=\"1.0\" lower=\"0.0\"/>
		</rateAG>
		<rateAT>
			<parameter id=\"at\" value=\"1.0\" lower=\"0.0\"/>
		</rateAT>
		<rateCG>
			<parameter id=\"cg\" value=\"1.0\" lower=\"0.0\"/>
		</rateCG>
		<rateGT>
			<parameter id=\"gt\" value=\"1.0\" lower=\"0.0\"/>
		</rateGT>
	</gtrModel>

	<!-- site model                                                              -->
	<siteModel id=\"siteModel\">
		<substitutionModel>
			<gtrModel idref=\"gtr\"/>
		</substitutionModel>
		<gammaShape gammaCategories=\"4\">
			<parameter id=\"alpha\" value=\"0.5\" lower=\"0.0\"/>
		</gammaShape>
		<proportionInvariant>
			<parameter id=\"pInv\" value=\"0.5\" lower=\"0.0\" upper=\"1.0\"/>
		</proportionInvariant>
	</siteModel>
	
		<treeLikelihood id=\"treeLikelihood\" useAmbiguities=\"false\">
		<patterns idref=\"patterns\"/>
		<treeModel idref=\"treeModel\"/>
		<siteModel idref=\"siteModel\"/>
		<discretizedBranchRates idref=\"branchRates\"/>
	</treeLikelihood>"

block.7 <- " <operators id=\"operators\" optimizationSchedule=\"default\">
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"ac\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"ag\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"at\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"cg\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"gt\"/>
		</scaleOperator>
		<deltaExchange delta=\"0.01\" weight=\"0.1\">
			<parameter idref=\"frequencies\"/>
		</deltaExchange>
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"alpha\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"0.1\">
			<parameter idref=\"pInv\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"3\">
			<parameter idref=\"ucld.mean\"/>
		</scaleOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"3\">
			<parameter idref=\"ucld.stdev\"/>
		</scaleOperator>
		<subtreeSlide size=\"86.0\" gaussian=\"true\" weight=\"15\">
			<treeModel idref=\"treeModel\"/>
		</subtreeSlide>
		<narrowExchange weight=\"15\">
			<treeModel idref=\"treeModel\"/>
		</narrowExchange>
		<wideExchange weight=\"3\">
			<treeModel idref=\"treeModel\"/>
		</wideExchange>
		<wilsonBalding weight=\"3\">
			<treeModel idref=\"treeModel\"/>
		</wilsonBalding>
		<scaleOperator scaleFactor=\"0.75\" weight=\"3\">
			<parameter idref=\"treeModel.rootHeight\"/>
		</scaleOperator>
		<uniformOperator weight=\"30\">
			<parameter idref=\"treeModel.internalNodeHeights\"/>
		</uniformOperator>
		<scaleOperator scaleFactor=\"0.75\" weight=\"3\">
			<parameter idref=\"constant.popSize\"/>
		</scaleOperator>
		<upDownOperator scaleFactor=\"0.75\" weight=\"3\">
			<up>
				<parameter idref=\"ucld.mean\"/>
			</up>
			<down>
				<parameter idref=\"treeModel.allInternalNodeHeights\"/>
			</down>
		</upDownOperator>
		<swapOperator size=\"1\" weight=\"10\" autoOptimize=\"false\">
			<parameter idref=\"branchRates.categories\"/>
		</swapOperator>
		<uniformIntegerOperator weight=\"10\">
			<parameter idref=\"branchRates.categories\"/>
		</uniformIntegerOperator>
	</operators>"


block.8 <-	"<mcmc id=\"mcmc\" chainLength=\"100000000\" autoOptimize=\"true\">
		<posterior id=\"posterior\">
			<prior id=\"prior\">
				<gammaPrior shape=\"0.05\" scale=\"10.0\" offset=\"0.0\">
					<parameter idref=\"ac\"/>
				</gammaPrior>
				<gammaPrior shape=\"0.05\" scale=\"20.0\" offset=\"0.0\">
					<parameter idref=\"ag\"/>
				</gammaPrior>
				<gammaPrior shape=\"0.05\" scale=\"10.0\" offset=\"0.0\">
					<parameter idref=\"at\"/>
				</gammaPrior>
				<gammaPrior shape=\"0.05\" scale=\"10.0\" offset=\"0.0\">
					<parameter idref=\"cg\"/>
				</gammaPrior>
				<gammaPrior shape=\"0.05\" scale=\"10.0\" offset=\"0.0\">
					<parameter idref=\"gt\"/>
				</gammaPrior>
				<uniformPrior lower=\"0.0\" upper=\"1.0\">
					<parameter idref=\"frequencies\"/>
				</uniformPrior>
				<exponentialPrior mean=\"0.5\" offset=\"0.0\">
					<parameter idref=\"alpha\"/>
				</exponentialPrior>
				<uniformPrior lower=\"0.0\" upper=\"1.0\">
					<parameter idref=\"pInv\"/>
				</uniformPrior>
				<exponentialPrior mean=\"0.3333333333333333\" offset=\"0.0\">
					<parameter idref=\"ucld.stdev\"/>
				</exponentialPrior>
				<uniformPrior lower=\"0.0\" upper=\"100.0\">
					<parameter idref=\"ucld.mean\"/>
				</uniformPrior>
				<oneOnXPrior>
					<parameter idref=\"constant.popSize\"/>
				</oneOnXPrior>
				<coalescentLikelihood idref=\"coalescent\"/>
			</prior>
			<likelihood id=\"likelihood\">
				<treeLikelihood idref=\"treeLikelihood\"/>
			</likelihood>
		</posterior>
		<operators idref=\"operators\"/>

		<!-- write log to screen                                                     -->
		<log id=\"screenLog\" logEvery=\"1000\">
			<column label=\"Posterior\" dp=\"4\" width=\"12\">
				<posterior idref=\"posterior\"/>
			</column>
			<column label=\"Prior\" dp=\"4\" width=\"12\">
				<prior idref=\"prior\"/>
			</column>
			<column label=\"Likelihood\" dp=\"4\" width=\"12\">
				<likelihood idref=\"likelihood\"/>
			</column>
			<column label=\"rootHeight\" sf=\"6\" width=\"12\">
				<parameter idref=\"treeModel.rootHeight\"/>
			</column>
			<column label=\"ucld.mean\" sf=\"6\" width=\"12\">
				<parameter idref=\"ucld.mean\"/>
			</column>
		</log>"
		
		
		

block.9 <- paste0("<log id=\"fileLog\" logEvery=\"1000\" fileName=\"", file.name, ".log\" overwrite=\"false\">")
block.9.1 <- "<posterior idref=\"posterior\"/>
			<prior idref=\"prior\"/>
			<likelihood idref=\"likelihood\"/>
			<parameter idref=\"treeModel.rootHeight\"/>
			<parameter idref=\"constant.popSize\"/>
			<parameter idref=\"ac\"/>
			<parameter idref=\"ag\"/>
			<parameter idref=\"at\"/>
			<parameter idref=\"cg\"/>
			<parameter idref=\"gt\"/>			
			<parameter idref=\"frequencies\"/>
			<parameter idref=\"alpha\"/>
			<parameter idref=\"pInv\"/>			
			<parameter idref=\"ucld.mean\"/>
			<parameter idref=\"ucld.stdev\"/>
			<rateStatistic idref=\"meanRate\"/>
			<rateStatistic idref=\"coefficientOfVariation\"/>
			<rateCovarianceStatistic idref=\"covariance\"/>
			<treeLikelihood idref=\"treeLikelihood\"/>
			<generalizedSkyLineLikelihood idref=\"coalescent\"/>
		</log>"
block.9 <- c(block.9, block.9.1)

block.10 <- paste0("<logTree id=\"treeFileLog\" logEvery=\"1000\" nexusFormat=\"true\" fileName=\"", file.name, ".trees\" sortTranslationTable=\"true\">")
block.10.1 <- "<treeModel idref=\"treeModel\"/>
			<trait name=\"rate\" tag=\"rate\">
				<discretizedBranchRates idref=\"branchRates\"/>
			</trait>
			<posterior idref=\"posterior\"/>
		</logTree>
	</mcmc>
	<report>
		<property name=\"timer\">
			<mcmc idref=\"mcmc\"/>
		</property>
	</report>
</beast>"

block.10 <- c(block.10, block.10.1)



if(xml.type == "random"){
    block.2 <- get.random.dates()
}else if(xml.type == "real"){
    block.2 <- get.real.dates()
}

all.blocks <- c(block.1, block.2, block.3, block.4, block.5, block.6, block.7, block.8, block.9, block.10)

writeLines(all.blocks, con = paste0(file.name, ".xml"))
}
